### ISSUE TITLE
Persisting Parallel State Machines Fails

### ISSUE THREAD

**Issue by ASKobayashi** [statelyai/xstate#871]
**Description**
I'm trying to repeatedly externally step a parallel state machine with persistence.  I have set up a simple linear machine that goes a -> b -> c -> d and notes entry and exit.  

The base machine is 
```
    states: {
      a: {
        entry: ["doEntry"],
        exit: ["doExit"],
        on: { TRIGGER: { target: "b" } }
      },
      b: {
        entry: ["doEntry"],
        exit: ["doExit"],
        on: { TRIGGER: { target: "c" } }
      },
      c: {
        entry: ["doEntry"],
        exit: ["doExit"],
        on: { TRIGGER: { target: "d" } }
      },
      d: { type: "final" }
    }
```

In the code sandbox I set up 4 tests.  
I create a parallel machine which uses two instance of the base machine.

1) Simple Linear works as expected.
2) Simple Parallel also works as expected.  You see two entrys to a, two exits of a.. etc
3) Step Linear Works, but complains about Warning: No implementation found for action type
     I would like to know why that is the case, but the output looks correct
4) Step Parallel does not work.

For quick reference, the serialize/deserialize code is snipped below:
```js
function step(machine, jsonState) {
  const service = interpret(machine);

  if (jsonState === null) {
    service.start();
  } else {
    var restoredState = State.create(JSON.parse(jsonState));
    service.start(restoredState);
    service.send("TRIGGER");
  }

  var retVal = JSON.stringify(service.state);
  service.stop();
  return retVal;
}

function parallelStepTest (machine) {
  console.log('start')
  jsonState.b = step(machine, null)
  console.log('step 1')
  jsonState.c = step(machine, jsonState.b)
  console.log('step 2')
  jsonState.d = step(machine, jsonState.c)
}
```

**Expected Result**
```
Step Parallel
start
entry a
entry a
step 1
exit a
exit a
entry b
entry b
step 2
exit b
exit b
entry c
entry c
step 3
exit c
exit c
```

**Actual Result**
```
Step Parallel
start
entry a
entry a
step 1
entry a
entry a
exit a
entry b
exit a
entry b
step 2
step 3
```

**Reproduction**

- https://codesandbox.io/s/xstate-parallel-machine-persist-cmh7k

**Additional context**

**Comment by Andarist** [comment]
Just a quick question so I get a better understanding of the reported issue - you want to rehydrate your child machines' states, right?

**Comment by ASKobayashi** [comment]
Hello @Andarist, thanks for following up. 

Yes, I would expect to be able to serialize the complex machine at any state and then resume it at a later point.  

Thanks for the tip there.  I didn't realize that capturing the state of a machine wouldn't include the child states.  I would expect it would do so.  Is there a case where you might want to save a parent state without capturing the child states?

Either way, I will investigate recursively descending into child properties to save/restore state.  If you or anyone else has an example of this, I would appreciate it very much.

Thanks again!

**Comment by ASKobayashi** [comment]
Has anyone successfully achieved this?  I'm having a lot of trouble restoring child states correctly.  

I created a new version of the code sandbox which captures the issue I'm having here:
https://codesandbox.io/s/xstate-parallel-machine-persist-ktjtw

It appears to correctly restore the state for step 1.. however on step 2 and further it hits 
`console.log('no service for ' + name)` because the call to  `service.start(resolvedState)` does not create the service.children map.  

If anyone has a simple example of recursive machine re-hydration, or a suggestion as to why service.start isn't creating the children I'd greatly appreciate it.

Thanks.

**Comment by wilson51492** [comment]
@ASKobayashi any luck on solving this?  I am running into a similar need.

So far, it seems like the current code may actually make it impossible, as the childMachine in the parent state is overwritten by the `exec` function https://github.com/davidkpiano/xstate/blob/master/packages/core/src/interpreter.ts#L871.

For this to work, I think the resolvedState would need to be called as the argument here
https://github.com/davidkpiano/xstate/blob/master/packages/core/src/interpreter.ts#L979

**Comment by ASKobayashi** [comment]
@wilson51492 Unfortunately, I don't have a solution yet.  I did do some digging after I saw  #941.  It looks like there's some upcoming work that may help address some of these issues..  I also noticed #164 demonstrates a somewhat similar task.  I haven't been able to take the time to revisit this yet and see if a fresh take will produce better results.  I'd love to hear if you're successful at this though!

**Comment by wilson51492** [comment]
@ASKobayashi thanks for the quick response.  I will let you know if I come up with anything.

**Comment by davidkpiano** [comment]
I'll take a look at this soon.

**Comment by ASKobayashi** [comment]
@davidkpiano I/We appreciate it.  XState is impressive work.  I would think wanting to persist complex states (and store them in a database) is a probably pretty common use case. Maybe if the documentation added an example of the "right way"™ to take a complex machine and persist/rehydrate, it would go a long way.  

As an example, the one at the bottom of https://xstate.js.org/docs/guides/parallel.html has hierarchy and parallel machines

Either way, thanks for creating XState.  It's impressive work.

**Comment by wilson51492** [comment]
dito to all of that ^

**Comment by davidkpiano** [comment]
I appreciate it, thank you! Persisting complex states should "just work" ideally.

**Comment by gsbelarus** [comment]
Any updates on how to persist/restore whole state of complex machine, including states of children sub machines?

**Comment by Ontopic** [comment]
Would love to hear the recommended approach as well.

**Comment by davidkpiano** [comment]
I have to write official docs on this, but you would want to use an **event sourcing** pattern: https://github.com/davidkpiano/xstate/discussions/1217#discussioncomment-23219

**Comment by lukenofurther** [comment]
@davidkpiano by event sourcing do you mean replaying each event in the machine's history on top of the starting state? 

The use case I'm building for will have hundreds and probably into the thousands of events with many side-effects to calculate subsequent contexts, as well as dozens of sub machines in a complex hierarchy. I feel an event sourcing pattern, if I'm understanding it correctly, would be very slow for such a machine once the history builds up.

**Comment by davidkpiano** [comment]
> The use case I'm building for will have hundreds and probably into the thousands of events with many side-effects to calculate subsequent contexts, as well as dozens of sub machines in a complex hierarchy. I feel an event sourcing pattern, if I'm understanding it correctly, would be very slow for such a machine once the history builds up.

This is a good concern; a lot of event-sourcing tools today have snapshotting for that reason, where instead of replaying e.g., 1000 events, you would replay a snapshot of the first 900 events + the remaining 100 events.

Still trying to figure this out, and because of how actors are being simplified in v5, it's likely that this will be a v5 feature.

**Comment by lukenofurther** [comment]
Snapshotting, that's an interesting idea.

I'm persisting the state in the `onTransition` event, so if there was an easy way to hook into that event on the child services as well, then we'd at least be able to persist the children manually. Similarly, restoring state occurs when starting the services, so an interface for accessing the child services would help there as well.

I'm looking into using the existing `invoke` interface to achieve something similar, for example directly providing an instance of `Interpreter` to invoke a child service. If that is possible, then the above would be doable already. I'll be trying that and a few other things out in the next wee while.

**Comment by hanzlahabib** [comment]
> I have to write official docs on this, but you would want to use an **event sourcing** pattern: [#1217 (comment)](https://github.com/davidkpiano/xstate/discussions/1217#discussioncomment-23219)

Thanks David, Event sourcing fails when you have some details you persisted in context through user journey. it may be solved if we save complete event object which contains type and its payload content for transition, there also happens to scenario where we don't want to persist some specific event payloads e.g global error messages which are handled from context and they are removed after number of seconds, if we use event sourcing technique it will populate context again with the error payload.
also i notice i often encounter multiple same events logging even after checking state.changed condition

**Comment by davidkpiano** [comment]
> can you brief difference in between (from xstate internal perspectives)
> `state.changed === true` & `state.changed !== undefined`

- `state.changed === undefined` signifies the initial state.
- `state.changed === true` signifies any changed state after the initial state.

**Comment by davidkpiano** [comment]
This should be fixed in v5 beta in #3743 - please try it out and let us know if it doesn't work as expected.
